pin/pin.go

```diff
@@ -211,22 +211,23 @@ func NewPinner(dstore ds.Datastore, serv, internal ipld.DAGService) Pinner {
 
 // Pin the given node, optionally recursive
 func (p *pinner) Pin(ctx context.Context, node ipld.Node, recurse bool) error {
-	p.lock.Lock()
-	defer p.lock.Unlock()
 	err := p.dserv.Add(ctx, node)
 	if err != nil {
 		return err
 	}
 
 	c := node.Cid()
 
+	p.lock.Lock()
```

Comment:
This seemed like an excessive locking to me. Unless I'm missing something, no need to lock the pin lists while writing content to the datastore.

Reply:
I agree. There was a lot of confusion about which locks had to be held at any given time for a while (we were using the pin lock to guard things that should have been guarded by the GC lock).